src/shillelagh/adapters/api/google_analytics.py

```diff
@@ -0,0 +1,218 @@
+
+class GoogleAnalytics(Adapter):
+    columns: Dict[str, Field] = {}
+
+    def __init__(
+        self,
+        uri,
+        service_account_info: str = None,
+        view_id=None,
+        operation: str = None,
+        parameters=Optional[Tuple[Any, ...]],
+        date_range = None,
+    ):
+        super().__init__()
+        self.view_id = view_id
+        self.operation = operation
+        self.parameters = parameters
+        self.date_range = date_range
+
+        credentials = ServiceAccountCredentials.from_json_keyfile_dict(
+            service_account_info, SCOPES
+        )
+        self.analytics_service = build(
+            "analyticsreporting", "v4", credentials=credentials
+        )
+
+        self.parsed_operation = parse(operation)
+        parsed_columns = self.get_query_columns()
+        self._set_columns(parsed_columns)
+
+    def supports(uri: str, fast: bool, **kwargs):
+        if uri.startswith("google_analytics"):
+            return True
+
+    @staticmethod
+    def parse_uri(uri: str) -> Tuple[Any, ...]:
+        return (uri,)
+
+    @staticmethod
+    def need_operation():
+        return True
+
+    def get_query_columns(self):
+        stmt = self.parsed_operation[0]
+        columns = []
+        column_identifiers = []
+
+        # get column_identifieres
+        in_select = False
+        for token in stmt.tokens:
+            if isinstance(token, Comment):
+                continue
+            if str(token).lower() == "select":
+                in_select = True
+            elif in_select and token.ttype is None:
+                if isinstance(token, Identifier):
+                    column_identifiers.append(token)
+                elif isinstance(token, IdentifierList):
+                    for identifier in token.get_identifiers():
+                        column_identifiers.append(identifier)
+                break
+
+        # get column names
+        for column_identifier in column_identifiers:
+            columns.append(column_identifier.get_name().replace('`', ''))
+
+        return columns
+
+    # def get_date_range(self):
+    #     parsed_operation = self.parsed_operation[0]
+    #     for token in parsed_operation:
+    #         if isinstance(token, Where) and token.value.startswith("where dateRange in"):
+    #             range_string = token.value.replace("where dateRange in", "").strip().replace("(", "").replace(")", "").split(",")
+    #             if (len(range_string) == 2):
+    #                 return range_string[0].strip(), range_string[1].strip()
+    #     return None
+
+    def get_data(
+        self, bound: Dict[str, Filter], order: List[Tuple[str, RequestedOrder]]
+    ) -> Iterator[Row]:
+        metrics = []
+        dimensions = []
+
+        for col in self.columns:
+            if col in ALL_DIMENSIONS:
+                dimensions.append({"name": col})
+            elif col in ALL_METRICS:
+                metrics.append({"expression": col})
```

Comment:
The adapter should always return all possible columns available, and let SQLite narrow it down to the columns it needs.

Reply:
hi @betodealmeida , there are few restrictions on GA apis so getting all dimenstions and columns isn't really possible with 1 call.
in 1 call only 9 dimensions and metrics, so do you suggest doing multiple calls dividing all possible dimensions and metrics?

Reply:
How about giving user a choice to select category while configuring.
A category would be 1 of the item on page -> https://ga-dev-tools.web.app/dimensions-metrics-explorer/
So then we can query all dimensions and metrics (or make multiple calls if they are more than 10).
but then user can only use 1 dataset on superset for 1 category which is bit inconvenient.

Reply:
@mayurnewase one solution might be creating virtual tables for each category. For example, this query:

```sql
SELECT * FROM "ga://User" WHERE "ga:userType" = 'New Visitor';
```

Would return the columns (dimensions and metrics) for "User" (`ga:userType`, `ga:sessionCount`, etc.). This would solve problem (2), but not (1).

For (1) I think the only option is either doing multiple calls or passing the list of requested columns. SQLite has a mechanism for passing the requested columns, but it's not exposed in `apsw` (I just left a comment on a [ticket](https://github.com/rogerbinns/apsw/issues/278) asking for it).

Would it be hard to implement multiple requests for now? Is there a common ID that you can use to join the requests?

Reply:
The 2nd problem here is if we wanna get all dimenstions and metrics of single category like user (if we do select * from ga://user) we still need to do some kind of batching logic as we can query columns and dimenstions in certain combo.
For example we can only query user_type dimension with Users, newUsers, percentNewSessions and sessionPerUser in an api call, and can't use any other dimension/metric in User category.

![Screenshot from 2022-08-07 17-40-39](https://user-images.githubusercontent.com/12967587/183290054-0fde630a-0959-4e5b-a4db-d5fced68a208.png)

To find such combos, you can try selecting a check box.